server/src/main/java/com/skillmagnet/User/UserController.java

```diff
@@ -107,4 +120,56 @@ public ResponseEntity<User> loginUser(@RequestBody User unAuthorizedUser) {
         userRepository.save(user);
         return new ResponseEntity<>(user, HttpStatus.OK);
     }
+    
+    /**
+     * This endpoint is to be used when a user "finishes" a lesson
+     * If requested lesson exists it's added to the users set of completed lessons
+     * 
+     * Tracking lessons completed allows us to be able to update course progress
+     * in enrollment objects which is also completed here automatically
+     * 
+     * @param uid - user that finished course
+     * @param lid - lesson they completed
+     * @return
+     *         Success: User Object & OK
+     *         Failure: null & NOT_FOUND
+     */
+    @PostMapping("/user/lesson/{uid}/{lid}")
+    public ResponseEntity<User> addLessonCompleted(@PathVariable("uid") int uid, @PathVariable("lid") int lid){
+        // If either lesson or user doesn't exist, return not found
+        User user = userRepository.findById(uid);
+        Optional<Lesson> lessonToAdd = lessonRepository.findById(lid);
+        if(lessonToAdd.isEmpty() || user == null){
+            return new ResponseEntity<>(null, HttpStatus.NOT_FOUND);
+        }
+
+        // Safe to assume that the Optional Contains a lesson at this point
+        // so we can use .get() knowing it will be a lesson added to the set
+        Set<Lesson> lessonsCompleted = user.getLessonsCompleted();
+        lessonsCompleted.add(lessonToAdd.get());
+        user.setLessonsCompleted(lessonsCompleted);
+
+        // Find enrollment object to update progress
+        Enrolls userEnrollment = enrollsRepository
+                .findByEnrolledCourseAndEnrolledUser(
+                    lessonToAdd.get().getCourse(), // Course
+                    user); // user
+        if(userEnrollment == null){
+            // A more detailed error probably needs to be here
+            // This would be an issue if a user is completing a lesson in a course
+            // they are not enrolled in
+            return new ResponseEntity<>(null, HttpStatus.NOT_FOUND);
```

Comment:
This situation comes up if client gives bad data so:

```suggestion
            return new ResponseEntity<>(null, HttpStatus.BAD_REQUEST);
```